packages/plugin-ext/src/plugin/file-system-ext-impl.ts

```diff
@@ -0,0 +1,396 @@
+        return this._proxy.$copy(source, destination, { ...{ overwrite: false }, ...options }).catch(ConsumerFileSystem._handleError);
+    }
+    private static _handleError(err: any): never {
+        // generic error
+        if (!(err instanceof Error)) {
+            throw new FileSystemError(String(err));
+        }
+
+        // no provider (unknown scheme) error
+        if (err.name === 'ENOPRO') {
+            throw FileSystemError.Unavailable(err.message);
+        }
+
+        // file system error
+        switch (err.name) {
+            case files.FileSystemProviderErrorCode.FileExists: throw FileSystemError.FileExists(err.message);
+            case files.FileSystemProviderErrorCode.FileNotFound: throw FileSystemError.FileNotFound(err.message);
+            case files.FileSystemProviderErrorCode.FileNotADirectory: throw FileSystemError.FileNotADirectory(err.message);
+            case files.FileSystemProviderErrorCode.FileIsADirectory: throw FileSystemError.FileIsADirectory(err.message);
+            case files.FileSystemProviderErrorCode.NoPermissions: throw FileSystemError.NoPermissions(err.message);
+            case files.FileSystemProviderErrorCode.Unavailable: throw FileSystemError.Unavailable(err.message);
+
+            default: throw new FileSystemError(err.message, err.name as files.FileSystemProviderErrorCode);
+        }
+    }
+}
+
+export class FileSystemExtImpl implements FileSystemExt {
+
+    private readonly _proxy: FileSystemMain;
+    private readonly _linkProvider = new FsLinkProvider();
+    private readonly _fsProvider = new Map<number, vscode.FileSystemProvider>();
+    private readonly _usedSchemes = new Set<string>();
+    private readonly _watches = new Map<number, IDisposable>();
+
+    private _linkProviderRegistration?: IDisposable;
+    private _handlePool: number = 0;
+
+    readonly fileSystem: vscode.FileSystem;
+
+    constructor(rpc: RPCProtocol, private _extHostLanguageFeatures: LanguagesExtImpl) {
+        this._proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.FILE_SYSTEM_MAIN);
+        this.fileSystem = new ConsumerFileSystem(this._proxy);
+
+        // register used schemes
+        Object.keys(Schemas).forEach(scheme => this._usedSchemes.add(scheme));
+    }
+
+    dispose(): void {
+        if (this._linkProviderRegistration) {
+            this._linkProviderRegistration.dispose();
+        }
+    }
+
+    private _registerLinkProviderIfNotYetRegistered(): void {
+        if (!this._linkProviderRegistration) {
+            this._linkProviderRegistration = this._extHostLanguageFeatures.registerDocumentLinkProvider('*', this._linkProvider, {
+                id: 'theia.fs-ext-impl',
+                name: 'fs-ext-impl'
+            });
+        }
+    }
+
+    registerFileSystemProvider(scheme: string, provider: vscode.FileSystemProvider, options: { isCaseSensitive?: boolean, isReadonly?: boolean } = {}) {
+
+        if (this._usedSchemes.has(scheme)) {
```

Comment:
I'm not sure this will work in the case of Che: since we're running plugin API in multiple hosts for the same front end, we'll end up running multiple instances of FileSystemExtImpl. Doing the uniqueness check on the ext-side will not work in this case.
@benoitf what do you think?

Reply:
Ok, seems that the FileService on the browser side checks for uniqueness on the schemes as well. So it's just an early exit, so to speak. Still it leads to two different error cases for the same error, so I would simply remove this check and rely on the one in the browser.

Reply:
This code is copied from VS Code repo without changes to align with VS Code extension expectations.